SignalServiceKit/src/Messages/OWSReadReceiptManager.m

```diff
@@ -122,51 +145,246 @@ - (void)process
+
         self.isProcessing = NO;
 
-        NSArray<OWSReadReceipt *> *readReceiptsToSend = [self.toLinkedDevicesReadReceiptMap allValues];
+        NSArray<OWSReadReceipt *> *readReceiptsForLinkedDevices = [self.toLinkedDevicesReadReceiptMap allValues];
         [self.toLinkedDevicesReadReceiptMap removeAllObjects];
-        if (readReceiptsToSend.count > 0) {
-            OWSReadReceiptsMessage *message = [[OWSReadReceiptsMessage alloc] initWithReadReceipts:readReceiptsToSend];
+        if (readReceiptsForLinkedDevices.count > 0) {
+            OWSReadReceiptsForLinkedDevicesMessage *message =
+                [[OWSReadReceiptsForLinkedDevicesMessage alloc] initWithReadReceipts:readReceiptsForLinkedDevices];
 
             dispatch_async(dispatch_get_main_queue(), ^{
                 [self.messageSender sendMessage:message
                     success:^{
                         DDLogInfo(@"%@ Successfully sent %zd read receipt to linked devices.",
                             self.tag,
-                            readReceiptsToSend.count);
+                            readReceiptsForLinkedDevices.count);
                     }
                     failure:^(NSError *error) {
                         DDLogError(@"%@ Failed to send read receipt to linked devices with error: %@", self.tag, error);
                     }];
             });
         }
+
+        NSArray<OWSReadReceipt *> *readReceiptsToSend = [[self.toLinkedDevicesReadReceiptMap allValues] copy];
+        [self.toLinkedDevicesReadReceiptMap removeAllObjects];
+        if (self.toSenderReadReceiptMap.count > 0) {
+            for (NSString *recipientId in self.toSenderReadReceiptMap) {
+                NSArray<NSNumber *> *timestamps = self.toSenderReadReceiptMap[recipientId];
+                OWSAssert(timestamps.count > 0);
+
+                TSThread *thread = [TSContactThread getOrCreateThreadWithContactId:recipientId];
+                OWSReadReceiptsForSenderMessage *message =
+                    [[OWSReadReceiptsForSenderMessage alloc] initWithThread:thread messageTimestamps:timestamps];
+
+                dispatch_async(dispatch_get_main_queue(), ^{
+                    [self.messageSender sendMessage:message
+                        success:^{
+                            DDLogInfo(@"%@ Successfully sent %zd read receipts to sender.",
+                                self.tag,
+                                readReceiptsToSend.count);
+                        }
+                        failure:^(NSError *error) {
+                            DDLogError(@"%@ Failed to send read receipts to sender with error: %@", self.tag, error);
+                        }];
+                });
+            }
+            [self.toSenderReadReceiptMap removeAllObjects];
+        }
     }
 }
 
-- (void)messageWasReadLocally:(TSIncomingMessage *)message;
+#pragma mark - Mark as Read Locally
+
+- (void)markAsReadLocallyBeforeTimestamp:(uint64_t)timestamp thread:(TSThread *)thread
```

Comment:
Pulled this method in here as a first step towards consolidating/normalizing the behavior for "read locally" and "read on linked device".